lib/Lower/Bridge.cpp

```diff
@@ -574,41 +380,328 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+    }
+    for (size_t i = 0, n = incrementLoopInfo.size(); i < n; ++i) {
+      genFIRIncrementLoopBegin(incrementLoopInfo[i]);
+    }
 
-  void genFIR(const Fortran::parser::AssociateConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::BlockConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::ChangeTeamConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::CriticalConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::DoConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::IfConstruct &) { TODO(); }
+    // Generate loop body code.
+    for (auto &e : *eval.evaluationList) {
+      genFIR(e, unstructuredContext);
+    }
+
+    // Generate end loop code.
+    if (infiniteLoop || whileCondition) {
+      genFIRUnconditionalBranch(doStmtEval.localBlocks[0]);
+    } else {
+      for (size_t i = incrementLoopInfo.size(); i > 0;) {
+        genFIRIncrementLoopEnd(incrementLoopInfo[--i]);
+      }
+    }
+  }
 
-  void genFIR(const SelectCaseConstruct &) { TODO(); }
-  void genFIR(const SelectRankConstruct &) { TODO(); }
-  void genFIR(const SelectTypeConstruct &) { TODO(); }
+  /// Generate FIR to begin a structured or unstructured increment loop.
+  void genFIRIncrementLoopBegin(IncrementLoopInfo &info) {
+    mlir::Location location = toLocation();
+    mlir::Type type = info.isStructured()
+                          ? mlir::IndexType::get(builder->getContext())
+                          : info.loopVariableType;
+    mlir::Value lowerValue = genFIRLoopIndex(info.lowerExpr, type);
+    mlir::Value upperValue = genFIRLoopIndex(info.upperExpr, type);
+    if (info.stepExpr.has_value()) {
+      info.stepValue = genFIRLoopIndex(*info.stepExpr, type);
+    }
+    info.loopVariable = createTemporary(location, *info.loopVariableSym);
+
+    // Structured loop - generate fir.loop.
+    if (info.isStructured()) {
+      info.insertionPoint = builder->saveInsertionPoint();
+      llvm::SmallVector<mlir::Value, 1> stepVector;
+      if (info.stepValue) {
+        stepVector.push_back(info.stepValue);
+      }
+      info.doLoop = builder->create<fir::LoopOp>(location, lowerValue,
+                                                 upperValue, stepVector);
+      builder->setInsertionPointToStart(info.doLoop.getBody());
+      mlir::Value shadow = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.getInductionVar());
+      localSymbols.pushShadowSymbol(*info.loopVariableSym, shadow);
+      return;
+    }
 
-  void genFIR(const Fortran::parser::WhereConstruct &) { TODO(); }
+    // Unstructured loop preheader code - initialize tripVariable, loopVariable.
+    if (!info.stepValue) {
+      info.stepValue = builder->create<mlir::ConstantOp>(
+          location, builder->getIntegerAttr(info.loopVariableType, 1));
+    }
+    mlir::Value tripCount;
+    tripCount = builder->create<mlir::SubIOp>(location, upperValue, lowerValue);
```

Comment:
This could be merged with the previous line.

Reply:
Yes, that's better.